File: client/site_tests/platform_MiniJailUidGid/platform_MiniJailUidGid.py
# Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import logging
import os
import re

from autotest_lib.client.bin import test
from autotest_lib.client.common_lib import error, utils

class platform_MiniJailUidGid(test.test):
    version = 1
    preserve_srcdir = True

    def setup(self):
        os.chdir(self.srcdir)
        utils.system('make clean')
        utils.system('make all')


    def __run_cmd(self, cmd):
        result = utils.system_output(cmd, retain_output=True,
                                     ignore_status=True)
        return result


    def run_once(self):
        # Check that --uid [number] works
        # @TODO(fes): The autotest framework seems to preserve the ownership
        # from the source, so thest tests fail unless the bindir is changed to
        # be owned by root but read/execute by anyone
        self.__run_cmd(('chown -R root:root ' + self.bindir));
        self.__run_cmd(('chmod 755 ' + self.bindir));
        check_cmd = (os.path.join(self.bindir, 'platform_MiniJailUidGid') +
              ' --checkUid=1000')
        cmd = ('/sbin/minijail --uid=1000 -- ' + check_cmd)
        result = self.__run_cmd(cmd)
        succeed_pattern = re.compile(r"SUCCEED: (.+)")
        success = succeed_pattern.findall(result)
        if len(success) == 0:
          raise error.TestFail('Set user id failed.')

        # Check that --gid [number] works
        check_cmd = (os.path.join(self.bindir, 'platform_MiniJailUidGid') +
              ' --checkGid=1000')
        cmd = ('/sbin/minijail --gid=1000 -- ' + check_cmd)
        result = self.__run_cmd(cmd)
        succeed_pattern = re.compile(r"SUCCEED: (.+)")
        success = succeed_pattern.findall(result)
        if len(success) == 0:
          raise error.TestFail('Set group id failed.')

File: client/site_tests/platform_MiniJailUidGid/src/platform_MiniJailUidGid.cc
// Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <errno.h>
#include <stdio.h>
#include <unistd.h>

#include "base/command_line.h"

// Verifies that the process group id is what is expected.
void CheckGid(uid_t gid) {
  uid_t rgid;
  uid_t egid;
  uid_t sgid;
  if (getresgid(&rgid, &egid, &sgid) != 0) {
    printf("ERROR: call to getresgid() failed: %d\n", errno);
  }
  if (rgid == gid && egid == gid && sgid == gid) {
    printf("SUCCEED: Real, Effective, and Saved Group IDs are %d\n", gid);
  } else {
    printf("FAIL: Group IDs: Real %d, Effective %d, Saved %d (Expected %d)\n",
           rgid, egid, sgid, gid);
  }
}

// Verifies that the process user id is what is expected.
void CheckUid(uid_t uid) {
  uid_t ruid;
  uid_t euid;
  uid_t suid;
  if (getresuid(&ruid, &euid, &suid) != 0) {
    printf("ERROR: call to getresuid() failed: %d\n", errno);
  }
  if (ruid == uid && euid == uid && suid == uid) {
    printf("SUCCEED: Real, Effective, and Saved User IDs are %d\n", uid);
  } else {
    printf("FAIL: User IDs: Real %d, Effective %d, Saved %d (Expected %d)\n",
           ruid, euid, suid, uid);
  }
}

int main(int argc, const char* argv[]) {
  CommandLine::Init(argc, argv);

  const CommandLine* cmd_line = CommandLine::ForCurrentProcess();

  if(cmd_line->HasSwitch("checkUid")) {
    int uid = atoi(cmd_line->GetSwitchValueASCII("checkUid").c_str());
    CheckUid(static_cast<uid_t>(uid));
  }

  if(cmd_line->HasSwitch("checkGid")) {
    int gid = atoi(cmd_line->GetSwitchValueASCII("checkGid").c_str());
    CheckGid(static_cast<uid_t>(gid));
  }

  return 0;
}
